lib/Cpdf.php

```diff
@@ -2329,6 +2621,13 @@ function output($debug = false)
         $content = '%PDF-1.7';
         $pos = mb_strlen($content, '8bit');
 
+        // pre-process o_font objects before output of all objects
```

Comment:
Related to your comment https://github.com/dompdf/dompdf/pull/2107#issuecomment-626379294
I worked on this after my previous comment and my first idea was also to move it to o_font out action directly.

I decided to take this little extra effort to keep the add step of o_font. The code modifies and adds new objects.
While it should be safe to assume that all other related objects are processed after o_font it is safer and clearer this way.

Reply:
My original thought was to process the full font without subsetting, then perform the subsetting during "out" (so "add" would have still been part of it). That way the full font information would be available right up until the PDF is generated. That would probably require a bit of extra processing but I didn't really have a chance to plan out that change fully.

Your approach isn't too far off from that, so I think it's good. I only have two potential concerns, mainly because I haven't fully analyzed the change yet:

1. Are there any internal objects that rely on the font information that is, with this change, now being generated only when the PDF is output?
2. Is there any reason a user of the library would be using that font information that will run into issues with this change?

If we're comfortable with the answer to those questions them I think this will work.

Reply:
Answers to your questions:

1. You are correct we need to analyze if any of the later steps is needed in CPdf. 
I have not too many doubts because OpenFont is still called and the font object is almost not changed in the following lines. From a quicker look just some information is added which is needed in the PDF and it also builds for example the fontDescriptor object. But I absolutely did not check this intensively enough.
2. Very much relates to 1. The main objects that are changing later are just the objects added to the pdf. The user should not want to read them but might write it which could be overwritten potentially.

I think it does not change much if we would double run the major part of this code block to have the font data in advance. If a user would interact with them he would read or write inconsistent data.
The main focus is on the data coming from openFont function and this is still processed early.

Reply:
I think we're OK with internal use. For external ... I guess we'll see.